Refactor storePodsNamespacer.List()

mml opened the pull request:
This fixes a bug in the previous version where, when we fell back on a
brute force approach, we were still returning an error.

It also clarifies the flow control into 3 distinct cases.  The cases
don't share variables any more, which makes mistakes like the one
mentioned above harder.

lavalamp commented:
code change looks like a big improvement. Should I be asking for a new test?

mml commented:
@lavalamp what do you have in mind?

lavalamp commented:
@mml maybe of the brute force list block, since I think it probably wasn't getting used/exercised at all before?

mml commented:
@lavalamp test added, and I verified that it failed on upstream/master.  PTAL

mml commented:
Huh.

```
Running tests for APIVersion: v1,extensions/v1beta1 with etcdPrefix: kubernetes.io/registry
+++ [0519 16:06:04] Running tests without code coverage
[...]
FAIL	k8s.io/kubernetes/test/integration	406.207s
```

mml commented:
@k8s-bot retest this please #25539

mml commented:
@k8s-bot test this issue: #25539

mml commented:
Well, this is a new one.

```
--- FAIL: TestReflectorResync (0.01s)
	reflector_test.go:380: exactly 2 iterations were expected, got: 3
[...]
FAIL
FAIL	k8s.io/kubernetes/pkg/client/cache	0.884s
```

Passes on my workstation.  I'm running it in a loop overnight hoping for a repeat failure.

mml commented:
I ran this 1000 times last night with 0 failures.

@k8s-bot unit test this issue: #25964

kargakis commented:
storeReplicationControllersNamespacer seems to have the same problem

mml commented:
@lavalamp OK PTAL.  I recommend looking at each commit separately.  I will squash once the review is done.

In the end, I realized upon looking at the way storeReplicationControllerNamespacer.List was implemented that the simple control flow works pretty easily as long as you never touch the return values until you get into one of the three cases.  Both functions are exactly the same now.

I think I uncovered yet another bug, though, which I'll file a new issue about.

mml committed:
Refactor *Namespacer.List().

Refactor storePodsNamespacer.List() and
storeReplicationContollersNamespacer.List().  They are the same
function, just with different signatures.

This fixes a bug where, when we fell back on a brute force approach, we
were still returning an error.

Also change to explicit return without named return values.

mml commented:
Updated again to fix #26107

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -77,37 +77,36 @@ type storePodsNamespacer struct {
 	key := &api.Pod{ObjectMeta: api.ObjectMeta{Namespace: s.namespace}}
 	items, err := s.indexer.Index(NamespaceIndex, key)
 	if err != nil {
+		err = nil
```
I am not sure this is optimal. At the very least, you should add a comment explaining why this needs to happen. Unfortunately named return values don't help much either.

mml replied:
Done.

lavalamp commented on pkg/client/cache/listers.go:
```diff
@@ -77,37 +77,37 @@ type storePodsNamespacer struct {
 			}
 		}
-		return list, nil
+		return
```
`return pods, err`

I don't care myself, but some people find it less confusing.

mml replied:
I'll make the change.

Would be nice to have a clear style guide and enforce it with a verify-* check.

lavalamp commented on pkg/client/cache/listers.go:
```diff
@@ -77,37 +77,37 @@ type storePodsNamespacer struct {
 			}
 		}
-		return list, err
+		return
```
same for all

lavalamp commented:
lgtm-- squash & add label yourself.
